xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,218 @@
+import io.grpc.xds.internal.rbac.engine.cel.DescriptorMessageProvider;
+import io.grpc.xds.internal.rbac.engine.cel.Dispatcher;
+import io.grpc.xds.internal.rbac.engine.cel.IncompleteData;
+import io.grpc.xds.internal.rbac.engine.cel.Interpretable;
+import io.grpc.xds.internal.rbac.engine.cel.Interpreter;
+import io.grpc.xds.internal.rbac.engine.cel.InterpreterException;
+import io.grpc.xds.internal.rbac.engine.cel.RuntimeTypeProvider;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * CEL Evaluation Engine is part of the authorization framework in gRPC.
+ * CEL Evaluation Engine takes one or two Envoy RBAC policies as input
+ * and uses CEL library to evaluate the condition field 
+ * inside each RBAC policy based on the provided Envoy Attributes. 
+ * CEL Evaluation Engine will generate an authorization decision which
+ * could be ALLOW, DENY or UNKNOWN.
+ * 
+ * <p>Use as in:
+ * 
+ * <pre>
+ *  AuthorizationEngine engine = new AuthorizationEngine(rbacPolicies);
+ *  AuthorizationDecision result = engine.evaluate();
+ * </pre>
+ */
+public class AuthorizationEngine<ReqT, RespT> {
+  private static final Logger log = Logger.getLogger(AuthorizationEngine.class.getName());
+  
+  /**
+   * RbacEngine is an inner class that holds RBAC action 
+   * and conditions of RBAC policy.
+   */
+  private static class RbacEngine {
+    @SuppressWarnings("UnusedVariable")
+    private final Action action;
+    private final ImmutableMap<String, Expr> conditions;
+
+    public RbacEngine(Action action, ImmutableMap<String, Expr> conditions) {
+      this.action = action;
+      this.conditions = conditions;
+    }
+  }
+
+  private final RbacEngine allowEngine;
+  private final RbacEngine denyEngine;
+
+  /**
+   * Creates a CEL-based Authorization Engine from a list of Envoy RBACs.
+   * The constructor can take either one or two Envoy RBACs.
+   * When it takes two RBACs, the order has to be a RBAC with DENY action
+   * followed by a RBAC with ALLOW action.
+   * @param rbacPolicies input Envoy RBAC list.
+   * @throws IllegalArgumentException if the user inputs an invalid RBAC list.
+   */
+  public AuthorizationEngine(ImmutableList<RBAC> rbacPolicies) throws IllegalArgumentException {
+    if (rbacPolicies.size() < 1 || rbacPolicies.size() > 2) {
+      throw new IllegalArgumentException(
+        "Invalid RBAC list size, must provide either one RBAC or two RBACs. ");
+    } 
+    if (rbacPolicies.size() == 2 && (rbacPolicies.get(0).getAction() != Action.DENY 
+        || rbacPolicies.get(1).getAction() != Action.ALLOW)) {
+      throw new IllegalArgumentException( "Invalid RBAC list, " 
+          + "must provide a RBAC with DENY action followed by a RBAC with ALLOW action. ");
+    }
+    RbacEngine allowEngine = null;
+    RbacEngine denyEngine = null;
```

Comment:
Since `this.allowEngine` is a final variable, am I still able to change the value later if I initialized it to null?

Reply:
Java doesn't allow us to assign final variables in a loop, so I guess we still have to do it the old way :(

Reply:
Instead of the list (and the corresponding loop) if you accept 2 params in the constructor you'll be able to do without the local vars and the loop. Of course this is not a serious issue but might point towards the simpler way of having 2 params

Reply:
Yes, that's a good point! Just want to double check with my host to see if we need to stay consistent with the design doc since it uses a list there :)

Reply:
Does the design doc mention why a list is used?